### DISCUSSION TITLE
FastAPI Testing and mocking

### DISCUSSION THREAD

**Discussion by naheedroomy** [fastapi/fastapi#4082] [Questions] [answered]
Hello!
New to FastAPI development, and I have a simple CRUD application for a user registration endpoint(signup, login, get users etc) and I've developed this with MongoDB and FastAPI.

I want to write some tests, pytests to be precise. However, I dont want my tests to be connecting to a database, and instead I'd like to use a fake database of some sorts, something like mongomock.

However there's no proper documentation on using mongomock for a fastapi project or how it works/how to integrate it.

Is there anyone who has any experience in this matter and would like to guide me?

For example, here's what my create user code looks like. 

```
@user.post('/register',
           response_model=User,
           responses={404: {"model": Message}},
           tags=["user"])
async def create_user(user: User):
    checkUserExists = MongoUser.objects.filter(email=user.email)

    if not checkUserExists:
        hashed = hashlib.sha256(user.password.encode('utf-8'))
        hashedPassword = hashed.hexdigest()
        MongoUser(email=user.email,
                  password=hashedPassword).save()
        MongoPasswordHistory(email=user.email,
                             password=hashedPassword).save()

        return JSONResponse(
            status_code=200,
            content={"message": "User created successfully"}
        )

    else:
        return JSONResponse(
            status_code=403,
            content={"message": "User already exists"}
        )`

MongoUser and MongoPasswordHistory are 2 tables in my database. 

The test I've written using the FastAPI documentation: 

```from fastapi.testclient import TestClient

def test_get_user_email_success():
    response = client.get("/users?email=example%40gmail.com")
    assert response.status_code == 200
    assert response.json() == "{"message": "User created successfully"}"
```
    
    
How do I do this without defining a database connection for the tests?Because right now when I run the tests, it creates entries on my DB.

**Comment by AlexandreBiguet** [comment, marked as answer]
Hi :) 

I guess you are using mongoengine (which I don't really know) and use its `connect` function which creates a session with the database so that you can start talking to the DB without an explicit "Session" object (or something like that ?).

From the documentation, there is a [register_connection function](https://docs.mongoengine.org/apireference.html#mongoengine.register_connection), that has a  `is_mock` parameter. It might be possible to register a mock connection at the beginning of the test file, which will override the default one defined in your app file ?

---

#### Not directly related, but perhaps useful ? 

If it is acceptable for you to use an explicit connection to the database, you could use dependency override feature, see [the documentation](https://fastapi.tiangolo.com/advanced/testing-dependencies/#testing-dependencies-with-overrides). The following snippet should by pytest-able (not  tested)

```python
import pymongo

from fastapi import FastAPI, Depends

app = FastAPI()


async def get_prod_client():
    return pymongo.MongoClient(host="mongodb://127.0.0.1:27017")


@app.get("/user/{email}")
async def user_get(email: str, client: pymongo.MongoClient = Depends(get_prod_client)):
    db = client["someDBName"]
    collection = db["users"]
    found_user = collection.find_one(filter={"email": email})
    return {"name": found_user["name"], "email": found_user["email"]}


## Test starts here.

import mongomock

from fastapi.testclient import TestClient


test_client = mongomock.MongoClient()


async def get_test_client():
    return test_client


app.dependency_overrides[get_prod_client] = get_test_client


class TestUser:
    def test_user_create_old(self):
        db = test_client["someDBName"]
        collection = db["users"]
        user = {"name": "Jane Doe", "email": "jdoe@example.com"}
        collection.insert_one(user)

        # Asserting inserted user is accessible from the DB

        found_user = collection.find_one(filter={"email": "jdoe@example.com"})
        assert user["name"] == found_user["name"]

        # Asserting inserted user is accessible from the API

        client = TestClient(app)
        response = client.get("/user/jdoe@example.com")
        response.raise_for_status()

        assert user["name"] == response.json()["name"]
``` 

Hope this helps, Cheers :)

**Reply by abvincita** [reply]
Sorry for the late reply, but did anyone ever get this working?

I followed your above example and it throws this error on the `response = client.get("/user/jdoe@example.com")` line:

> ValueError: [TypeError("'ObjectId' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]

Still struggling to get my FastAPI with mongodb testable using mongomock and dependency override :(
And I'm not even using mongoengine or motor, just straight pymongo as per @AlexandreBiguet's example.

**Reply by callamd** [reply]
Basically unless you want to use dependency injection for everything you are S.O.L. Thanks FastAPI.

**Reply by n8sty** [reply]
That's not entirely true. It's very possible using one of the objects from `unittest.mock` or the `pytest` fixture `monkeypatch` to mock or monkeypatch however connections to Mongo are being created. So, while the above shows overriding FastAPI style dependencies, there should be other ways that aren't specific to FastAPI.
